lib/events/filesessions/fileops.go

```diff
@@ -0,0 +1,108 @@
+package filesessions
+
+import (
+	"context"
+	"io"
+	"log/slog"
+	"os"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	reservationFilePerm = 0600
+	combinedFilePerm    = reservationFilePerm
+)
+
+// FileOps captures file operations done by filesessions
+type FileOps interface {
+	CreateReservation(ctx context.Context, name string, size int64) error
+	WriteReservation(ctx context.Context, name string, data io.Reader) error
+	CombineParts(ctx context.Context, dst io.Writer, parts []string) error
+}
+
+func loggingClose(ctx context.Context, closer io.Closer, log *slog.Logger, msg string, args ...any) {
+	if err := closer.Close(); err != nil {
+		log.ErrorContext(ctx, msg, append(args, "error", err)...)
+	}
+}
+
+type plainFileOps struct {
+	log      *slog.Logger
+	openFile utils.OpenFileWithFlagsFunc
+}
+
+var _ FileOps = &plainFileOps{}
+
+// NewPlainFileOps returns a plaintext implementation of the FileOps interface.
+func NewPlainFileOps(log *slog.Logger, openFile utils.OpenFileWithFlagsFunc) *plainFileOps {
+	return &plainFileOps{log: log, openFile: openFile}
+}
+
+// CreateReservation creates the initial reservation file for
+func (p *plainFileOps) CreateReservation(ctx context.Context, name string, size int64) (err error) {
+	defer func() {
+		if err != nil {
+			err = trace.ConvertSystemError(err)
+		}
+	}()
+
+	f, err := p.openFile(name, os.O_WRONLY|os.O_CREATE, reservationFilePerm)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := f.Truncate(size); err != nil {
+		loggingClose(ctx, f, p.log, "failed to close file after Truncate error", "name", name)
+	}
+
+	return trace.Wrap(f.Close())
+}
+
+func (p *plainFileOps) WriteReservation(ctx context.Context, name string, data io.Reader) (err error) {
+	defer func() {
+		if err != nil {
+			err = trace.ConvertSystemError(err)
+		}
+	}()
+
+	log := p.log.With("name", name)
+	// O_CREATE kepr for backwards compatibility only
+	const flag = os.O_WRONLY | os.O_CREATE
+
+	f, err := p.openFile(name, flag, reservationFilePerm)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer loggingClose(ctx, f, log, "failed to close reservation file during write")
+
+	n, err := io.Copy(f, data)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := f.Truncate(n); err != nil {
```

Comment:
if any of these operation fails, should we keep the file?

Reply:
Probably not. The caller of `WritePart()` removes the file in case of a failure and there's no reason to hang onto the reserved file if we bail out because of a failed truncation.

The individual part files should probably be removed too, but we don't currently remove them on failure (at least not near the callsite) and I don't really want to change effective behavior with this PR